optuna/integration/botorch.py

```diff
@@ -488,8 +459,33 @@ def sample_relative(
                     value *= -1
                 values[trial_idx, obj_idx] = value
 
-            if con is not None:
-                con[trial_idx] = self._trial_constraints[trial_idx]
+            if self._constraints_func is not None:
+                constraints = study._storage.get_trial_system_attrs(trial._trial_id)[
+                    "botorch:constraints"
+                ]
+                if constraints is not None:
+                    n_constraints = len(constraints)
+
+                    if con is None:
+                        con = numpy.full((n_trials, n_constraints), numpy.nan, dtype=numpy.float64)
+                    elif n_constraints != con.shape[1]:
+                        raise RuntimeError(
+                            f"Expected {con.shape[1]} constraints but received {n_constraints}."
+                        )
+
+                    con[trial_idx] = constraints
+
+        if self._constraints_func is not None:
+            if con is None:
+                warnings.warn(
+                    "`constraints_func` was given but no call to it correctly computed "
+                    "constraints. Constraints passed to `candidates_func` will be `None`."
+                )
+            elif numpy.isnan(con).any():
+                warnings.warn(
+                    "`constraints_func` was given but some calls to it did not correctly compute "
+                    "constraints. Constraints passed to `candidates_func` will contain NaN."
+                )
```

Comment:
It seems that there is no test for this case. How about adding it?

Reply:
Sure, let me add tests to check the elements for NaN as well.

Reply:
Added a test for the warning.